### DISCUSSION TITLE
BE Spring Bean 토론 | 2022.04.29

### DISCUSSION THREAD

**Discussion by woowabrie** [woowacourse/retrospective#13] [General]
## '어떤 객체를 스프링이 관리하게 해야 할까?' 에 대해 토론한 내용을 남겨주세요.
- 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
- 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
- (+ 시간이 된다면) '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.

## 끝장토론 대진표
![image](https://user-images.githubusercontent.com/46308949/165894448-3781052b-8cc4-4936-8c88-0e1981242b6f.png)

## 체크사항
✅ '어떤 생각과 고민의 과정을 거쳤는 지'가 잘 드러날 수 있게 이야기를 나누고 글로도 적어주세요.

**Comment by YJGwon** [comment]
## 짱십 & 알렉스 & 페퍼 & 포키 끝장토론

- 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
    - 페퍼: Controller, Service, Dao
    - 일동: 만장일치
    - 안건 종료
- 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
    - 포키: 스프링은 웹 어플리케이션을 만들기 위한 프레임 워크니까 웹 구동을 위한 객체들을 등록하면 되지 않을까?
    - 일동: 만장일치
    - 안건 종료
    - 짱십: 뒷자리에서 컨텍스트 실행 시점에 객체가 되어야 하면 빈이라는 이야기를 한다
    - 포키: 킹치만 요청 scope 주기에서만 사용하게 설정할수도 있는걸..?
    - 일동: 근데 컨텍스트 실행 시점이 뭐지?
    - 일동: 밥 먹으면서 알아보자. (배고프다는 뜻)

**Reply by Ohzzi** [reply]
다 먹고 살자고 하는 짓인데 밥 먹어야죠

**Reply by sc0116** [reply]
@Ohzzi  💩

**Reply by SuyeonChoi** [reply]
😱😶‍🌫️😋

**Reply by bugoverdose** [reply]
dto가 아니라 dao아닌가여

**Reply by YJGwon** [reply]
포차차~~

**Comment by jurlring** [comment]
## 돌범 & 주디 & 쬬그린 & 후니의 끝장 토론

- 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
돌: 컨트롤러, 서비스, 레포, password encoder
후: 컨트롤러, 서비스, 레포, password encoder
쬬: 컨트롤러, 서비스, 레포
주:컨트롤러, 서비스, 레포

- 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.

컨트롤러 서비스 레포 공통점

1. 레이어를 대표한다.
2. 프로그램에서 하나만 올려서 사용한다.
3. 기본 빈 스코프 전략이 싱글톤 전략이라면 프로그램 시작과 종료로 life cycle을 관리할 수 있다.
4. 빈 장점 ⇒ 스프링 컨테이너에서 싱글톤으로 관리되므로 리소스 이득
- (+ 시간이 된다면) '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.

application 단위로 spring이 관리해주는 특별한 싱글톤을 사용할 수 있다. ⇒ 서버 성능 향상에 도움이 된다.

**Reply by jojogreen91** [reply]
주디👍

**Reply by jayjaehunchoi** [reply]
와우 ;; 대박

**Reply by is2js** [reply]
ㅎㅎ;;

**Comment by hamcheeseburger** [comment]
# 루나, 쿼리치, 이스트, 코린

### 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
- Controller, Service, Dao 객체들만 스프링 빈으로 생성했습니다. 

### 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
- 빈으로 등록된 객체가 가지는 필드들은 모두 빈으로 등록되어야 한다고 생각합니다.
- 도메인이 특정 객체의 상태가 된다면, 여러 사용자가 동시에 이용할 시에 상태 값이 변하게 되므로 원하는 결과 값을 얻을 수 없습니다. 따라서 도메인이 빈으로 등록된 객체의 상태가 되면 안됩니다. 그러므로 도메인은 빈으로 등록되면 안된다고 생각합니다.

### '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.
- 빈을 스프링이 관리하면 DI 간편하게 이루어지기 때문입니다.
- 싱글톤으로 객체를 관리하기 때문입니다.

**Comment by Byeongju-Kong** [comment]
오리, 티키, 라쿤, 크리스

### 어떤 객체를 빈으로 만들었는지
- Controller
- Service
- Dao

### 빈으로 만들 객체에 대한 기준
- 스프링이 관리를 해야하는 객체
- 매번 다른 객체를 생성하지 않아도 되는 객체

### 스프링이 객체를 관리하게 해야 하는 이유
- 의존성 주입을 스프링이 관리하도록 하기 위함
    - 의존성 주입을 스프링이 하게 하는 이유 : 개발자의 편의 (개발자가 개발하는 시점에 의존 관계에 대해 생각하지 않을 수 있다 )

**Comment by sinb57** [comment]
### 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?

- 이프 : Controller Service Repository Dao
- 이브 : Controller Service Dao
- 썬 : Controller Service Dao
- 봄 : Controller Service Dao


### 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.

- application 단위로 싱글톤 방식의 사용이 가능해야 한다.
- 상태를 지니면 안된다.(무상태성) 예를 들어 Service가 Game을 인스턴스로 지니면 안된다.



### '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.

- 객체 생성과 사용 로직을 분리할 수 있다.

**Comment by seong-wooo** [comment]
# 리버 마루 케이 배카라 

### 1. 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
컨트롤러, 서비스, dao

### 2. 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
딱 하나만 있어도 되는 객체를 스프링 싱글톤으로 구현한다.

### 3. '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.
요청이 동시에 여러개 들어올 때, 성능상의 이점을 위해서 스프링이 관리하게 한다.

**Comment by tonic523** [comment]
# 토닉, 레오, 슬로, 렉스 끝장토론

## 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
- 컨트롤러, 서비스, 레포지토리 관련 객체를 관리하게 했다.

## 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
- 싱글톤으로 관리해야 하는 객체
- 요청이 올 때마다 사용하는 객체와 의존되어 있는 객체

(밥먹으면서 조금 더 고민할 예정)

**Reply by YJGwon** [reply]
이 팀도 배고팠네

**Comment by hyeonic** [comment]
## 끝장 토론 (릭, 엘리, 매트)

- 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
    - `릭냥이`:
        - [https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-introduction](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-introduction)
        - (공식문서 중 일부) In Spring, the objects that form the `backbone of your application` and that are managed by the Spring IoC container are called beans. A bean is an object that is instantiated, assembled, and managed by a Spring IoC container. Otherwise, a bean is simply one of many objects in your application. Beans, and the dependencies among them, are reflected in the configuration metadata used by a container.
    - `엘리`: @Controller, @Service, @Repository에 사용했다. 해당 애노테이션을 가지고 스프링이 인식해서 기능을 제공하는 것으로 이해했다. 내 프로젝트에서 스프링의 기능을 적용할 수 있는 부분이 어디있을지 고민한 뒤 접근했다.
    - `매트`: 도메인 객체 처럼 상태를 가진 객체는 적합하지 않다고 생각을 했고, `공통적으로 사용할 수 있는 행위`들을 가진 객체가 스프링 빈으로 사용할 때 적합하다고 생각한다.
- 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
    - `릭냥이`: 스프링 MVC가 웹 서비스를 구현할 때 프레임워크라고 생각했다. 이러한 웹 서비스에서 추상적으로 비슷한 구조들이 그려진다. 그런 역할을 하는 객체를 빈으로 등록해야 한다.
    - `엘리`: 모든 객체를 빈으로 등록해야 하는 것은 아니다. 스프링 컨테이너가 여러 가지 기능을 제공해준다. 그러한 기능을 맡길 필요가 없으면 굳이 등록하지 않는다. (그러한 기능?)
    - `매트`: 우선 애플리케이션 전반에서 기반이 되는 객체 들을 등록해야 한다고 생각한다. 이 객체는 여러 객체에 주입되어 사용될 수 있기 때문에 변하는 상태를 가지면 안된다.
- (+ 시간이 된다면) '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.
    - 스프링 컨테이너가 제공하는 기능(IoC, DI 등)들을 적절히 사용하기 위해서!
    - 개발자가 관리해야 할 포인트들을 줄일 수 있다.
    - 개발자가 비즈니스에 집중할 수 있다.

**Reply by her0807** [reply]
릭냥이 ㅋㅋ

**Comment by kbsat** [comment]
# 로마, 소주캉, 에덴, 아서 끝장토론

### 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?

- 컨트롤러, 서비스 Dao 정도를 만들었다.

### 스프링 빈으로 만들 객체에 대한 기준

- 상태가 없는 객체 : 사이드 이펙트가 다른 곳으로 전파될 수 있다.
- 의존성 관리가 되어야하는 객체
- 객체의 생성이나 소멸주기를 신경써주지 않아도 되는 객체

### 왜 객체를 스프링이 관리해야할까 ? 왜 빈을 사용해야하는가?

- 빈을 사용하면 관리의 용이함. 불필요하거나 귀찮은 작업을 스프링에서 해주고 개발자가 핵심 비즈니스 로직에만 집중하도록 도와준다.

**Comment by tco0427** [comment]
# 디우, 애쉬, 클레이, 승팡

---

## 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
- Service, Repository, Controller를 스프링 빈으로 등록하였다.
    -> Service, Repository, Controller를 빈으로 등록함으로써 재사용성을 높이고, 의존성 주입을 스프링에게 맡긴다.

## 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
- 스프링 빈으로 만들 객체에 대한 기준
    -> 상태를 가지는 객체는 빈으로 등록하지 않는다.
    -> 왜? 빈으로 등록하면 싱글톤으로 관리되는데 여러 클라이언트에서 상태를 공유하게 된다.

* 어플리케이션에서 사용할 인터페이스의 구현체가 빈으로 등록되어야 한다.

* 스프링은 빈을 싱글톤으로 관리하니까 상태가 없는 객체를 빈으로 등록한다. 

* 어플리케이션 실행시점에 결정되는 것은 빈으로 등록하고, 실행시간(특정 시점)에 객체가 생성되고 소멸되는 경우는 빈으로 등록하지 않는다.

**Comment by soominsohn** [comment]
**야호 & 판다 & 헌치 & 앤지**
---


- 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
    - Serivce, Dao, Controller에 붙임
- 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
    - 주입하는 객체가 싱글톤 일 때 (모든 구조가 불변일 때)
    - 불변이 아닌 경우? 그 때 그 때 생성하는 것보다 한번에 등록하고 관리하는 객체
    - Service, Dao, Controller등에만 붙임
    - 어떤 객체를 스프링이 관리해야하나?
        - 도메인을 제외한 객체
        - 스프링으로 웹 띄우고 요청 받는 객체
    - dto는 스프링으로 관리하나요?
        - ModelAttribute 사용하려고 감싸려고 dto를 만듬
        - dto는 어노테이션을 안붙였는데 알아서 찾아간다.
        - 왜 그럴까?
        - 어노테이션을 안붙이면 빈으로 관리가 안되는 것일까?
    - domain 에서는 가져올 때 정상적으로 작동을 한다.
        - bean으로 등록하는 주체는 service, dao, repository, service 등이 있다.
    - 왜 스프링이 도메인을 관리해주면 안되는걸까?
        - 싱글톤 이야기
        - 도메인은 인스턴스를 생성하기 때문에 싱글톤이 아니다.
        - 싱글톤이 아닌 객체들은 동기화를 하는데도 비용이 든다.
        - Color와 같은 enum식의 도메인은 빈으로 만들어줘도 괜찮지 않을까?
    - Spring에서 모든걸 bean으로 만드는 것이 필요할까?
        - 장점이 있을까?
        - 도메인으로 연결되는 객체인 경우 하나하나 bean으로 등록할 필요는 없지 않나?
- (+ 시간이 된다면) '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.
    - 어노테이션을 붙이면 객체를 어디서 생성하고 실행하는 것을 스프링이 알아서 해준다.
        - 자원적으로 계속 객체를 생성하는 것에 비해 좋다.

**Comment by Yboyu0u** [comment]
## 파잉후아의 토론. (파랑, 잉, 후디, 아스피)
### 어떤 객체를 스프링이 관리하게 해야할까?
- (잉) 빈으로 등록하면 장점 : new 키워드로 생성할 필요 없다.

- (파랑) 큰 고민을 아직 해보지 않았다. 자동으로 등록해주고, 생성하거나 관리할 필요 없이 스프링에서 알아서 싱글톤으로 관리해주는 것이 편했다.

- (아스피) 스프링에서 빈에 등록해놓으면, 의존성 주입을 알아서 해주고, 우리는 비즈니스 로직에만 집중할 수 있다. 도메인 객체는 왜 빈으로 등록안할까? => 그에 대한 기준은 예를 들어 체스방을 여러개 생성할 수 있고, ... 이런것들은 싱글톤으로 관리되면 안된다. Service같은 것은 내부의 변도잉 생길일이 없으니까 그런 친구들은 등록해두고, 싱글톤으로 관리하면 되지 않을까. 라는 나름의 기준을 정했다.

- (후디) 의문점이 하나 생기는데, 어떤 설정을 해주면 빈을 싱글턴이 아니라 멀티턴으로 관리할 수 있다고 해줬는데 이에 대한 생각이 궁금.

- (아스피) 멀티톤으로 관리할 수 있다고 해도, 도메인 객체를 빈으로 관리하는 것이 불가능하지 않을까?

- (잉) 관리해야할 것이 많아지면 그때 시점부터는 빈으로 관리하는 것이 좋지 않을까?

- (아스피) DataSource 를 빈으로 등록해놓은 상태인데, 그것을 사용하려면 Dao 를 쓰려면 빈으로 등록하고, 서비스를 사용하려면 Dao .. 다 물고물고 연결되어 빈으로 등록해져야 한다.

- (아스피) DataSource 를 생성하는 건 스프링에 맡기자.

- (후디) 스프링에 비즈니스 로직 외의 관심사를 맡기는 것이 좋다고

- (후디) 비즈니스 로직에 관련된 객체는 프로그래머가 제어의 권한을 가지고, 그외의 것들은 빈으로 등록해서 스프링에 제어의 권한을 주는 것


유익한 토론이였네요🙏🏻

**Reply by YJGwon** [reply]
아임파잉땡큐앤유~?

**Comment by her0807** [comment]
## 스컬/오찌 + 쿤/수달 조 내용

### 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
`Controller, Service, Dao`

### 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.

웹을  동작하게 하기 위해 필요한 최소한의 객체만 빈으로 등록한다. 
MVC 레이어에 속한 객체들은 최소한 스프링에서 관리해줘야한다. 
스프링 라이프 사이클과 일치하는 객체들을 스프링에서 관리해줘야한다.
일시적으로 요청받아서 생성, 사용하는 객체들은 관리할 필요 없음 (ex dto, domain.. ) 

### 왜 객체를 스프링이 관리하게 해야 할까?
- 스프링을 사용하여 객체가 싱글톤일 경우 대규모 트래픽일 때 객체 생성이 적어져 트래픽 관리에 용이함
- 스프링을 사용하면, 간결하게 중요한 로직에만 집중할 수 있게 됨
- 스프링과 함께 사용할 수 있는 JPA , Security 등등 을 사용할 수 있어서 자동화 가능
- 스프링은 프레임워크다 -> 라이브러리에 정의된 코드를 가져다 쓰냐  vs 프레임워크가 우리 코드를 제공하냐의 차이 


### 그외 
 스프링 컨테이너에서 싱글톤으로 객체를 관리함. -> 왜? -> 싱글톤일 경우 대규모 트레픽일 때 객체 생성이 적어져 트레픽 관리에 용이함
여기서 말하는 싱글톤과, 그냥 싱글톤 패턴이랑 약간 다름 ( 스콥이라는 키워드로 찾아보기) 

IOC 에 대해 궁금함 -> 이게 과연 스프링에만 존재하는 이점인가?

**Comment by awesomeo184** [comment]
어떤 객체를 빈으로 등록해야할까?
* 스프링에 이미 등록된 빈을 객체가 사용해야 할 때
* 스프링이 제공하는 기능을 객체가 사용해야 할 때

객체를 스프링이 관리할때 장점이 무엇인가
* 스프링이 제공하는 기능을 사용할 수 있으니 편하다!
  * 싱글톤으로 관리해준다.
  * 의존성 주입(혹은 관리)이 편하다
  * 스프링에 이미 등록된 빈을 주입 받아 사용할 수 있다.

차리
* 어떤 객체를 빈으로 등록? : 컨트롤러, 서비스, 레포지토리, PasswordEncoder
* 우선 빈으로 등록하려면 상태가 없어야 한다.
* 작은 프로그램이라서 그런진 모르겠지만, 컨트롤러, 서비스, DAO 외에는 사용할 곳이 없는 것 같다.
* 도메인을 스프링 빈으로 등록했을 때, 얻을 수 있는 장점이 있을까?

어썸오
* 어떤 객체를 빈으로 등록? : 컨트롤러, 서비스, 레포지토리
* 상태를 가지고 있는데 빈으로 등록해야하는 상황이라면?
* 도메인에서 스프링이 제공하는 빈을 사용해야하는 상황이라면 어떻게 해야하나요?

제로
* 어떤 객체를 빈으로 등록? : 컨트롤러, 서비스, 레포지토리
* 프로토타입 스코프를 사용하면 상태를 가질 수 있는 것으로 앎

바니
* 어떤 객체를 빈으로 등록? : 컨트롤러, 서비스, 레포지토리
* 제로와 같은 의견, 공부를 더 해봐야함
* 싱글톤으로 관리되기 때문에 메모리 측면이나 성능에서 이득
* 의존성 주입을 편하게 해줌


'왜 객체를 스프링이 관리하게 해야 할까?'
* 밥먹으며 생각해보자

**Comment by bugoverdose** [comment]
## 아키, 연로그, 정, 토르

### **어떤 객체**를 스프링이 관리하도록 해야 할까? 
1. 스프링에게 의존성 주입 작업을 담당하도록 설정하고 싶은 객체들 
2. (기본적으로) 애플리케이션 내에서 하나의 객체만 생성되어 재활용되도록 설정하고 싶은 객체들 
3. (기본적으로) 상태를 지니지 않는 객체들

예시: 컨트롤러, 서비스, DAO 외에도 WebMvcConfigurer 등 스프링에서 제공하는 기능들을 구현하기 위한 객체들

### `@Bean`으로 빈을 등록해야 하는 경우는 혹시 없을까?

있다! 외부 라이브러리와 같이 클래스를 직접적으로 수정하기 힘든 경우, 이를 import해오고 인스턴스를 생성하여 직접 명시적으로 빈으로 등록하는 등 상황에 따라 필요할 수 있다.
- [참고 자료](https://jojoldu.tistory.com/27)

**Comment by ldk980130** [comment]
## 그린론, 범고래, 써머, 더즈

**체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?**
Controller, Service, Dao

**스프링 빈으로 만들면 스프링이 멀티 스레드에서 싱글톤이 보장이 된다.**
만드는 기준은 의존성 주입을 하는 경우에 싱글톤으로 관리되어야 할 때

**동시성 이슈?**
상태를 가지면 안 된다.
싱글톤으로 관리되기 때문에 상태를 가지는 객체를 빈으로 등록하면 동시성 이슈가 생길 수 있다.

**Contrller나 Service를 인스턴스로 만들면 안 되나?**
메모리 낭비를 일으킬 수 있음

**스프링이 객체를 관리하는 이점?**
의존성 관리를 주입 받고 했을 때 순환 참조나 싱글톤 관리를 잘 해준다.
변경이 덜하다. OCP를 지킬 수 있다. (DI를 통해 OCP를 지킬 수 있다.)
비즈니스 로직에 더 집중할 수 있다.

**DTO?**
DAO에서 뷰에스 쓰일 DTO를 반환하는것은 뷰의 변경이 DAO로 침투하는 것 같아 좋지 않다.
모든 요청에 DTO를 만들어야 하나?

**Reply by cndqjacndqja** [reply]
더즈 깃똥차네요!

**Comment by sure-why-not** [comment]
### 아리 & 유콩 & 알파 & 라라의 끝장 토론
**체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?**
- Service
- Controller
- Dao

**토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.**
- 상태를 갖지 않는 객체

**'왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.**
- 싱글톤으로 생성되기 때문에 성능이 좋아진다.
- 개발자가 집중해야할 범위가 줄어든다. 도메인에 집중하면 된다.

**Comment by yhh1056** [comment]
# 호호, 레넌, 베루스, 루키

## 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?

- 컨트롤러, 서비스, DAO클래스를 빈으로 생성했다.

## 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.

- 스프링 프레임워크가 사용해야 하는 빈으로 등록한다.
- 사용자가 호출할 때 필요한 객체들을 스프링 빈으로 등록해야 한다.
- 도메인을 빈으로 등록하면 안되나?
    - 도메인을 추가할 필요가 없다.
    - 스코프을 지정해서 사용할 수 있을 것 같지만 그렇게 사용해야 하나?
    - 도메인을 생성시점에 의존이 필요한가? 필요하지 않아서 빈으로 등록할 필요가 없다.
    - 도메인도 빈으로 등록하면 관리해야할 빈이 많아질 것 같다.
    - 도메인은 개발자가 관심을 가지고 만드는 객체다. 프레임워크에 의존할 필요가 없어서 등록할 필요가 없다고 생각한다.
    
    
- **These bean definitions correspond to the actual objects that make up your application. Typically, you define service layer objects, data access objects (DAOs), presentation objects such as Struts Action instances, infrastructure objects such as Hibernate SessionFactories, JMS Queues, and so forth. Typically, one does not configure fine-grained domain objects in the container, because it is usually the responsibility of DAOs and business logic to create and load domain objects.** (공식 문서에서 확인) [링크](https://docs.spring.io/spring-framework/docs/current/reference/html/core.html#beans-factory-metadata)

## + 시간이 된다면) '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.
- 개발자가 비즈니스 로직에 집중하기 위해서
- 스프링이 제공하는 의존성 주입을 편하게 하기 위해서

**Comment by progress0407** [comment]
# Spring Bean 토론 : 열음, 조시, 칙촉 , 필즈

## 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
    - 열음: controller, service, dao
    - 조시: controller, service, dao
    - 칙촉: contoller, service, dao
    - 필즈: controller, service, dao
    
    만장일치 !!
    
## 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
- 상태를 가지고 있지 않아야한다.
- 실행 시점에서 변화가 없는 객체인 경우
- 응답과 요청을 처리하는 객체인 경우

  

## (+ 시간이 된다면) '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.
    - 인터페이스 및에 구현체가 여러개 있을 경우 변경에 유리하다
    - 어느 곳이서든 객체를 주입받을 수 있다.
    - 오로지 하나만 관리될 객체  (물론 Bean Scope를 변경해서 생애주기를 교체할 수 있다)

## 크루 의견

### 열음
---

controller, service, dao를 bean으로 등록하여 사용했습니다.

빈으로 만들어서 스프링에 등록하고, 가져다 쓰게끔 하는 느낌으로 이해해서 

- 실행 시점에서 변화가 없는 객체인 경우
- 응답과 요청을 처리하는 객체인 경우

라는 기준으로 구분했습니다.

### 조시

---

- controller, service, dao에서 사용했습니다. 크루들이 이야기 하는걸 들어보니 스프링은 객체를 싱글톤으로 관리하기 때문에 필드를 사용할떄 주의해야 한다고 하더라고요. 여러 상태를 가지지 않는 무언가를 받고 처리해주는 객체일때 관리하면 좋지 않을까 생각해 봅니다!

### 칙촉

---

service, controller, dao 를 bean 으로 등록해서 사용했어요! 자동으로 의존성 주입을 해줘서 편하게 사용할 수 있어서? 아직 잘 모르겠네요.

### 필즈

---

#### 장점

1. 오로지 하나만 관리될 객체  (물론 Bean Scope를 변경해서 생애주기를 교체할 수 있다)
2. `Interceptor`, `AOP` 등 스프링이 지원하는 기술을 수혜받을 수 있다
3. `new` 로 생성될 수 있는 (일반적인) 객체이기 때문에 유틸리티 클래스와 달리 객체의 다형성, 상속 등의 객체 성질을 이용할 수 있다
4. 어느 곳이서든 객체를 주입받을 수 있다. 
5. `@Repository` 의 경우 예외가 나면 스프링이 추상화한 예외로 변환되는 기능이 있다
    - 즉 스프링이 가지고 있는 예외로 관리되어 일관성이 생김

#### 단점

1. 프레임워크에 종속적이게 된다.. 모든게 스프링의 컨테이너에 종속적이게 된다..

    - 객체를 관리하는 것 마저 스프링이 하게 되니.. 역으로 IoC 에 객체를 조작할 때는 힘들다.  

1. 테스트할 떄 조차 스프링의 문서를 읽으면서 어노테이션의 사용법을 익혀야 한다.

**Comment by OzRagwort** [comment]
# 리차드, 찬, 기론, 알린의 빈 토론

'어떤 객체를 스프링이 관리하게 해야 할까?' 에 대해 토론한 내용을 남겨주세요.
* 체스 미션에서 어떤 객체를 스프링 빈으로 만들었나요?
* 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.
* (+ 시간이 된다면) '왜 객체를 스프링이 관리하게 해야 할까?' 에 대해 논의하고 그 이유를 정리해주세요.
---

## 빈이 무엇일까? 왜 쓸까?

빈은 한마디로 @Component 가 달린 클래스들을 빈으로 관리한다.
빈으로 등록하는 이유는 객체지향적으로 코드를 만들기 위함이다.
빈으로 등록되면 애플리케이션의 생명주기를 스프링이 대신 해준다.
빈은 싱글톤으로 관리되어서 멀티쓰레드에서도 변경에 좀 더 안전하다.
개발자가 관리하기 어려운 부분을 많이 대신해준다.

## 어떤 객체를 스프링 빈으로 만들었나요?

@SpringBootApplication
@Controller
@Service
@Repository

spring-boot-starter-jdbc
JdbcTemplate
JdbcTemplate이 자동으로 빈주입이 되어있었다

BCryptPasswordEncoder
스프링 시큐리티에서 제공하는 단방향 암호화
@Configuration
@Bean

## 왜 이런 것을 빈으로 만들었지?

인터페이스의 구현체들이다
약한 의존을 만들기 위해서 구현체를 의존받는게 아닌 인터페이스를 의존받는데
이런 것들을 프로그램의 규모가 작으면 할 수 있어도 커지면 관리하기 어렵다.
사용할 의존체를 빈으로 만들어서 사용한다.

## 토론을 통해 나온 스프링 빈으로 만들 객체에 대한 기준을 설명해주세요.

의존성 주입을 개발자가 하지 않고 스프링이 해줘야 하니까 빈으로 등록한다! ==> 의존성 관리

Controller에서 시작하여 의존성 주입을 받는 객체들을 빈으로 등록하고 관리해야한다.
자신이 사용할 구현체를 빈으로 등록해야한다.
애플리케이션의 생명주기를 스프링에서 대신해줄 수 있는 장점이 있다.
객체지향적인 코드와 스프링 프레임워크의 장점을 살리기 위해 의존성을 관리할 빈을 등록한다.

## 왜 객체를 스프링이 관리하게 해야 할까?

예를들어 JdbcTemplate을 만드는 데에도 의존이 되있는 빈들이 20개가까이 되는데

이거 직접 가능?

애플리케이션의 규모가 커지면 커질수록 개발자가 의존성을 직접 관리하기 어렵다.

추가/수정/제거 모든 경우에도 어렵다

직접 관리를 해야한다면 모두 알아야하는데 이걸 기억하고 있기 어렵다.

**반박시 직접 구현해보세요**

# 초꿀팁

Spring 다이어그램 - 빈 종속성 에서 등록된 빈을 확인할 수 있다.
이 많은 빈을 개발자가 관리하기 어려울 것 같다.

<img width="1374" alt="image" src="https://user-images.githubusercontent.com/32123302/165883364-64c3b6fa-de07-4c3d-95ef-db834afc96c6.png">

-리차드

**Reply by Byeongju-Kong** [reply]
반박시 직접 구현해보세요ㅋㅋㅋㅋㅋ👍🏻
